rust/loqui_client/src/client.rs

```diff
@@ -0,0 +1,46 @@
+use crate::connection_handler::{ConnectionHandler, InternalEvent};
+use crate::Config;
+use failure::Error;
+use futures::sync::oneshot;
+use loqui_connection::{Encoder, LoquiError, Supervisor as SupervisedConnection};
+use std::net::SocketAddr;
+use std::sync::Arc;
+use tokio::await;
+
+#[derive(Clone)]
+pub struct Client<E: Encoder> {
+    connection: Arc<SupervisedConnection<ConnectionHandler<E>>>,
+}
+
+impl<E: Encoder> Client<E> {
+    pub async fn connect<A: AsRef<str>>(
+        address_str: A,
+        config: Config<E>,
+    ) -> Result<Client<E>, Error> {
+        let address: SocketAddr = address_str.as_ref().parse()?;
+        let config = Arc::new(config);
+        let handler_creator = move || ConnectionHandler::new(config.clone());
+        let connection = await!(SupervisedConnection::spawn(address, handler_creator));
+        let client = Self {
+            connection: Arc::new(connection),
+        };
+        Ok(client)
+    }
+
+    /// Send a request to the server.
+    pub async fn request(&self, payload: E::Encoded) -> Result<E::Decoded, Error> {
```

Comment:
General question: Should we use `failure::Error` here, or our own error enum? That way it's much more explicit to the caller about the kinda errors that can surface.

Reply:
I think we should use our own enum.